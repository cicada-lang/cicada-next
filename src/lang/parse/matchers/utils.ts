import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
