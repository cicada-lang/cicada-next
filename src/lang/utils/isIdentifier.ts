const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
