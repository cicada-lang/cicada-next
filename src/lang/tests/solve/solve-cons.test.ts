import { expect, test } from "vitest"
import { runCode } from "../utils"

test("solve Cons -- car", async () => {
  const output = await runCode(`

solve (a: String) {
  unify cons(a, "b") = cons("a", "b")
}

`)

  expect(output).toMatchInlineSnapshot('"{ a: \\"a\\" }"')
})

test("solve Cons -- cdr", async () => {
  const output = await runCode(`

solve (b: String) {
  unify cons("a", "b") = cons("a", b)
}

`)

  expect(output).toMatchInlineSnapshot('"{ b: \\"b\\" }"')
})

test("solve Cons -- car and cdr", async () => {
  const output = await runCode(`

solve (a: String, b: String) {
  unify cons(a, "b") = cons("a", b)
}

`)

  expect(output).toMatchInlineSnapshot('"{ a: \\"a\\", b: \\"b\\" }"')
})

test("solve Cons -- nested", async () => {
  const output = await runCode(`

solve (a: String, b: String) {
  unify cons(a, cons("b", "c")) = cons("a", cons(b, "c"))
}

`)

  expect(output).toMatchInlineSnapshot('"{ a: \\"a\\", b: \\"b\\" }"')
})
