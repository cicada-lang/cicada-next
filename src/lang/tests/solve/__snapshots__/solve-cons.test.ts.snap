// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solve Cons -- car > 
    "{
      a: \\"a\\"
    }"
   1`] = `
"{
  a: \\"a\\"
}"
`;

exports[`solve Cons -- car and cdr > 
    "{
      a: \\"a\\",
      b: \\"b\\"
    }"
   1`] = `
"{
  a: \\"a\\",
  b: \\"b\\"
}"
`;

exports[`solve Cons -- car and cdr: 
    "{
      a: "a",
      b: "b"
    }"
   1`] = `
"{
  a: "a",
  b: "b"
}"
`;

exports[`solve Cons -- car: 
    "{
      a: "a"
    }"
   1`] = `
"{
  a: "a"
}"
`;

exports[`solve Cons -- cdr > 
    "{
      b: \\"b\\"
    }"
   1`] = `
"{
  b: \\"b\\"
}"
`;

exports[`solve Cons -- cdr: 
    "{
      b: "b"
    }"
   1`] = `
"{
  b: "b"
}"
`;

exports[`solve Cons -- deepWalk > 
    "{
      a: \\"a\\",
      b: \\"b\\",
      c: cons(\\"a\\", \\"b\\")
    }"
   1`] = `
"{
  a: \\"a\\",
  b: \\"b\\",
  c: cons(\\"a\\", \\"b\\")
}"
`;

exports[`solve Cons -- deepWalk: 
    "{
      a: "a",
      b: "b",
      c: cons("a", "b")
    }"
   1`] = `
"{
  a: "a",
  b: "b",
  c: cons("a", "b")
}"
`;

exports[`solve Cons -- nested > 
    "{
      a: \\"a\\",
      b: \\"b\\"
    }"
   1`] = `
"{
  a: \\"a\\",
  b: \\"b\\"
}"
`;

exports[`solve Cons -- nested: 
    "{
      a: "a",
      b: "b"
    }"
   1`] = `
"{
  a: "a",
  b: "b"
}"
`;

exports[`solve Cons -- car 1`] = `
"{
  a: "a"
}"
`;

exports[`solve Cons -- cdr 1`] = `
"{
  b: "b"
}"
`;

exports[`solve Cons -- car and cdr 1`] = `
"{
  a: "a",
  b: "b"
}"
`;

exports[`solve Cons -- nested 1`] = `
"{
  a: "a",
  b: "b"
}"
`;

exports[`solve Cons -- deepWalk 1`] = `
"{
  a: "a",
  b: "b",
  c: cons("a", "b")
}"
`;
